presto-main/src/main/java/com/facebook/presto/operator/window/WindowPartition.java

```diff
@@ -194,19 +216,62 @@ else if (frameInfo.getType() == RANGE) {
         return new Range(frameStart, frameEnd);
     }
 
-    private boolean emptyFrame(FrameInfo frameInfo, int rowPosition, int endPosition)
+    private int precedingEndRange(long endValue)
+    {
+        int peerGroupEndIndex = peerGroupEndIndices.indexOf(peerGroupEnd);
```

Comment:
I'd use `Map<Integer, Integer>` (peer group -> position) instead of `ArrayList<Integer>`, to avoid slowness when there are many tiny groups in a large partition.
Imagine billions of rows with ~1 row / peer group.
Then `indexOf` will be O(billions). But if you use a Map, it's gonna be O(log(billions)).